parser-typechecker/src/Unison/Syntax/TermPrinter.hs

```diff
@@ -734,69 +778,80 @@ a + b = ...
+  m (Pretty SyntaxText)
+prettyBinding = prettyBinding0 $ ac (-1) Block Map.empty MaybeDoc
 
 prettyBinding' ::
-  Var v =>
+  MonadPretty v m =>
   Width ->
-  PrettyPrintEnv ->
   HQ.HashQualified Name ->
   Term v a ->
-  ColorText
-prettyBinding' width n v t =
-  PP.render width $ PP.syntaxToColor $ prettyBinding n v t
+  m ColorText
+prettyBinding' width v t =
+  PP.render width . PP.syntaxToColor <$> prettyBinding v t
 
 prettyBinding0 ::
-  Var v =>
-  PrettyPrintEnv ->
+  MonadPretty v m =>
   AmbientContext ->
   HQ.HashQualified Name ->
   Term2 v at ap v a ->
-  Pretty SyntaxText
-prettyBinding0 env a@AmbientContext {imports = im, docContext = doc} v term =
-  go
-    (symbolic && isBinary term)
-    term
+  m (Pretty SyntaxText)
+prettyBinding0 a@AmbientContext {imports = im, docContext = doc} v term =
+  go (symbolic && isBinary term) term
   where
-    go infix' = \case
-      Ann' tm tp ->
-        PP.lines
-          [ PP.group
-              ( renderName v
-                  <> PP.hang
-                    (fmt S.TypeAscriptionColon " :")
-                    (TypePrinter.pretty0 env im (-1) tp)
-              ),
-            PP.group (prettyBinding0 env a v tm)
-          ]
-      (printAnnotate env -> LamsNamedMatch' vs branches) ->
-        PP.group $
-          PP.group
-            ( defnLhs v vs <> fmt S.BindingEquals " =" <> " "
-                <> fmt
-                  S.ControlKeyword
-                  "cases"
-            )
-            `PP.hang` printCase env im doc branches
-      LamsNamedOrDelay' vs body ->
-        let (im', uses) = calcImports im body'
-            -- In the case where we're being called from inside `pretty0`, this
-            -- call to printAnnotate is unfortunately repeating work we've already
-            -- done.
-            body' = printAnnotate env body
-            -- Special case for 'let being on the same line
-            hang = case body' of
-              Delay' (Lets' _ _) -> PP.softHang
-              _ -> PP.hang
-         in PP.group $
-              PP.group (defnLhs v vs <> fmt S.BindingEquals " =")
-                `hang` uses [pretty0 env (ac (-1) Block im' doc) body']
-      t -> l "error: " <> l (show t)
+    go infix' binding = do
+      env <- getPPE
+      case binding of
+        Ann' tm tp -> do
+          -- If the term is an annotated function,
+          -- we want to print the type signature on the previous line.
+          -- The TypePrinter.pretty0 function prints the type, and uses a
+          -- Reader monad with (Set v) in it to track which type variables are
+          -- bound in the outer scope. We use that to determine if the type
+          -- printer should avoid capture of those variables.
+          let avoidCapture = case tp of
+                ForallsNamed' vs _ -> addTypeVars vs
+                _ -> id
+          tp' <- TypePrinter.pretty0 im (-1) tp
+          tm' <- avoidCapture (prettyBinding0 a v tm)
```

Comment:
This is the real meat of the fix. If the type binds type variables, they are bound in the scope of the lambda it annotates. I'm avoiding capture of those variables by adding them to the `local` scope. `addTypeVars` calls `local` which means these variables will go out of scope when the printer for the binding is finished.

Reply:
Yup, makes sense.